chalice/package.py

```diff
@@ -539,15 +566,290 @@ def _register_cfn_resource_name(self, name):
+
+        template['resource'].setdefault(
+            'aws_cloudwatch_event_rule', {})[
+                resource.resource_name] = {
+                    'name': resource.resource_name,
+                    'schedule_expression': resource.schedule_expression
+        }
+        template['resource'].setdefault(
+            'aws_cloudwatch_event_target', {})[
+                resource.resource_name] = {
+                    'rule': '${aws_cloudwatch_event_rule.%s.name}' % (
+                        resource.resource_name),
+                    'target_id': resource.resource_name,
+                    'arn': self._fref(resource.lambda_function)
+        }
+        template['resource'].setdefault(
+            'aws_lambda_permission', {})[
+                resource.resource_name] = {
+                    'function_name': resource.lambda_function.function_name,
+                    'action': 'lambda:InvokeFunction',
+                    'principal': 'events.amazonaws.com',
+                    'source_arn': "${aws_cloudwatch_event_rule.%s.arn}" % (
+                        resource.resource_name)
+        }
+
+    def _generate_lambdafunction(self, resource, template):
+        # type: (models.LambdaFunction, Dict[str, Any]) -> None
+
+        func_definition = {
+            'function_name': resource.function_name,
+            'runtime': resource.runtime,
+            'handler': resource.handler,
+            'memory_size': resource.memory_size,
+            'tags': resource.tags,
+            'timeout': resource.timeout,
+            'source_code_hash': '${filebase64sha256("%s")}' % (
+                resource.deployment_package.filename),
+            'filename': resource.deployment_package.filename}
+
+        if resource.security_group_ids and resource.subnet_ids:
+            func_definition['vpc_config'] = {
+                'subnet_ids': resource.subnet_ids,
+                'security_group_ids': resource.security_group_ids
+            }
+        if resource.reserved_concurrency is not None:
+            func_definition['reserved_concurrent_executions'] = (
+                resource.reserved_concurrency
+            )
+        if resource.environment_variables:
+            func_definition['environment'] = {
+                'variables': resource.environment_variables
+            }
+        if resource.layers:
+            func_definition['layers'] = list(resource.layers)
+
+        if isinstance(resource.role, models.ManagedIAMRole):
+            func_definition['role'] = '${aws_iam_role.%s.arn}' % (
+                resource.role.resource_name)
+        else:
+            # resource is a PreCreatedIAMRole.
+            role = cast(models.PreCreatedIAMRole, resource.role)
+            func_definition['role'] = role.role_arn
+
+        template['resource'].setdefault('aws_lambda_function', {})[
+            resource.resource_name] = func_definition
+
+    def _generate_restapi(self, resource, template):
```

Comment:
So I'm not sure what the issue is but I'm running into issues when redeploying. Specifically, if I start with the following `app.py`:
```python
from chalice import Chalice, Rate

app = Chalice(app_name='tf')
app.debug = True


@app.route('/')
def index():
    return {'hello': 'world'}
```
I then deployed it using:
```
$ chalice package --pkg-format=terraform out
$ cd out
$ terraform apply
```
And made a request to it and it worked:
```
$ http $(terraform output EndpointURL)
HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 17
Content-Type: application/json
Date: Sun, 04 Aug 2019 20:07:50 GMT
Via: 1.1 3cc149abcbd8cb50553c09666247bcfa.cloudfront.net (CloudFront)
X-Amz-Cf-Id: xaIvOdgvZbyNX43qQaZ2RHebbWASYpyO_gbgMkeKppBxufo6x_ZWjg==
X-Amz-Cf-Pop: HIO51-C1
X-Amzn-Trace-Id: Root=1-5d473b15-672b38e5fe6ae2f0e7875408;Sampled=0
X-Cache: Miss from cloudfront
x-amz-apigw-id: d6YrZGNOPHcFldQ=
x-amzn-RequestId: 88a2a7ab-b6f3-11e9-aa00-f7175e1fd48e

{
    "hello": "world"
}
```

Then I updated the `app.py` to and redeployed it:
```python
from chalice import Chalice, Rate

app = Chalice(app_name='tf')
app.debug = True


@app.route('/', methods=['POST'])
def index():
    return {'hello': 'world'}
```
And then I got a missing authentication token when I redeployed:
```
$ http POST $(terraform output EndpointURL)
HTTP/1.1 403 Forbidden
Connection: keep-alive
Content-Length: 42
Content-Type: application/json
Date: Sun, 04 Aug 2019 20:13:06 GMT
Via: 1.1 bb14b7a4f45e3e0b1ae5065c59f7826f.cloudfront.net (CloudFront)
X-Amz-Cf-Id: nxQo7BEVTn-D2Qff9UuTf3AmnwTUgnwJshmliHAgn3K5yD7yQNaQxQ==
X-Amz-Cf-Pop: HIO51-C1
X-Cache: Error from cloudfront
x-amz-apigw-id: d6Zc5GclPHcFUSA=
x-amzn-ErrorType: MissingAuthenticationTokenException
x-amzn-RequestId: 4576a6d9-b6f4-11e9-859c-8361244f785a

{
    "message": "Missing Authentication Token"
}
```

The weird part is that if I then run `terraform destroy` and then `terraform apply`, it then works:
```
$ http POST $(terraform output EndpointURL)
HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 17
Content-Type: application/json
Date: Sun, 04 Aug 2019 20:16:55 GMT
Via: 1.1 7e81687e34febf53ccb2929005d5ef12.cloudfront.net (CloudFront)
X-Amz-Cf-Id: zDynWia8aB3I1Fw0bYcD8BUdzfqgd8HucfZEIuWil-oPGxUe_01JsA==
X-Amz-Cf-Pop: HIO51-C1
X-Amzn-Trace-Id: Root=1-5d473d37-3831e3d2e82333a598e64473;Sampled=0
X-Cache: Miss from cloudfront
x-amz-apigw-id: d6aAmEiqPHcFazQ=
x-amzn-RequestId: cdae530f-b6f4-11e9-8e63-8f62322c2646

{
    "hello": "world"
}
```
Have you run into this issue before? I want to dive into why the API is not being setup correctly. I suspect it is a terraform bug though given it works after tearing it down and redeploying it.

Reply:
reproduced, the issue it appears is that terraform needs to create a new deployment on the method changes and doesn't because none of the deployment's attributes changed.

Reply:
resolved via removing explicit stage creation, and implicitly creating one via deployment, and forcing redeployments on swagger changes by linking in the checksum of the swagger.